pkg/storage/etcd/etcd_helper_test.go

```diff
@@ -535,3 +559,79 @@ func TestPrefixEtcdKey(t *testing.T) {
 
 	assert.Equal(t, keyBefore, keyAfter, "Prefix incorrectly added by EtcdHelper")
 }
+
+func TestDeleteUIDMismatch(t *testing.T) {
+	server := etcdtesting.NewEtcdTestClientServer(t)
+	defer server.Terminate(t)
+	prefix := path.Join("/", etcdtest.PathPrefix())
+	helper := newEtcdHelper(server.Client, testapi.Default.Codec(), prefix)
+
+	obj := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo", UID: "A"}}
+	podPtr := &api.Pod{}
+	err := helper.Create(context.TODO(), "/some/key", obj, podPtr, 0)
+	if err != nil {
+		t.Fatalf("Unexpected error %#v", err)
+	}
+	err = helper.Delete(context.TODO(), "/some/key", obj, api.NewPreconditionsWithUID("B"))
+	err = etcderr.InterpretDeleteError(err, unversioned.GroupResource{}, "")
+	if !errors.IsConflict(err) {
+		t.Fatalf("Expect a Conflict error, got: %v", err)
+	}
+}
+
+type getFunc func(ctx context.Context, key string, opts *etcd.GetOptions) (*etcd.Response, error)
+
+type FakeDeleteKeysAPI struct {
+	etcd.KeysAPI
+	FakeGetFunc getFunc
+	GetCount    *int
+	FakeGetCap  int
+}
+
+func (f FakeDeleteKeysAPI) Get(ctx context.Context, key string, opts *etcd.GetOptions) (*etcd.Response, error) {
+	*f.GetCount++
+	if *f.GetCount < f.FakeGetCap {
+		return f.FakeGetFunc(ctx, key, opts)
+	}
+	return f.KeysAPI.Get(ctx, key, opts)
+}
+
+// This is to emulate the case where another party updates the object when
+// etcdHelper.Delete has verified the preconditions, but hasn't carried out the
+// deletion yet. Etcd will fail the deletion and report the conflict. etcdHelper
+// should retry until there is no conflict.
+func TestDeleteWithRetry(t *testing.T) {
+	server := etcdtesting.NewEtcdTestClientServer(t)
+	defer server.Terminate(t)
+	prefix := path.Join("/", etcdtest.PathPrefix())
+
+	obj := &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo", UID: "A"}}
+	// fakeGet returns a large ModifiedIndex to emulate the case that another
+	// party has updated the object.
+	fakeGet := func(ctx context.Context, key string, opts *etcd.GetOptions) (*etcd.Response, error) {
+		data, _ := runtime.Encode(testapi.Default.Codec(), obj)
+		return &etcd.Response{Node: &etcd.Node{Value: string(data), ModifiedIndex: 99}}, nil
+	}
+	getCount := 0
+	expectedRetries := 3
+	helper := newEtcdHelper(server.Client, testapi.Default.Codec(), prefix)
+	helper.etcdKeysAPI = FakeDeleteKeysAPI{KeysAPI: helper.etcdKeysAPI, GetCount: &getCount, FakeGetCap: expectedRetries, FakeGetFunc: fakeGet}
```

Comment:
store a pointer to the fakeDeleteKeysAPI object instead of storing a pointer to getCount?